### ISSUE TITLE
Can't listen on port 53 with systemd on host

### ISSUE THREAD

**Issue by Malvineous** [moby/moby#32545]
**Description**

If the host is set up to use `systemd` for network configuration, then it is not possible to expose port 53 to a container.  Docker reports the port being in use when it is not and can be bound by other programs.

**Steps to reproduce the issue:**
1. `systemctl start systemd-resolved` (problem only exists when resolved is running)
2. `docker run -p 53:53 -p 53:53/udp ...`
3. See error `Error starting userland proxy: listen udp 0.0.0.0:53: bind: address already in use`
4. Use `netcat` to check ports: `nc -s 0.0.0.0 -l -p 53` (tcp) `nc -s 0.0.0.0 -l -u -p 53` (udp), `netcat` is able to listen on both TCP and UDP port 53 without problem, data comes in over the network and is visible in the netcat output.
5. `netstat -nlp` reports no program listening on port 53, only `systemd-resolved` listening on `127.0.0.53:53` which is not used by Docker.

**Describe the results you received:**

The container failed to run, claiming the port was in use.

**Describe the results you expected:**

The container should have run, as the port is not in use.

**Additional information you deem important (e.g. issue happens only occasionally):**

Stopping `systemd-resolved` before starting Docker works.

Instead of using `systemd-resolved`, binding port 53 on another interface will also stop the container from running, e.g:

    netcat -s 127.0.0.53 -l -u -p 53

Having this command running while the container is being launched will result in the same error complaining `0.0.0.0:53` is in use, when again, it is not in use.

**Output of `docker version`:**

```
Client:
 Version:      17.04.0-ce
 API version:  1.28
 Go version:   go1.8
 Git commit:   4845c567eb
 Built:        Sat Apr  8 18:55:45 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.04.0-ce
 API version:  1.28 (minimum version 1.12)
 Go version:   go1.8
 Git commit:   4845c567eb
 Built:        Sat Apr  8 18:55:45 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

**Output of `docker info`:**

```
Containers: 7
 Running: 1
 Paused: 0
 Stopped: 6
Images: 266
Server Version: 17.04.0-ce
Storage Driver: devicemapper
 Pool Name: docker-8:3-1794688-pool
 Pool Blocksize: 65.54kB
 Base Device Size: 10.74GB
 Backing Filesystem: xfs
 Data file: /dev/loop0
 Metadata file: /dev/loop1
 Data Space Used: 4.54GB
 Data Space Total: 107.4GB
 Data Space Available: 10.76GB
 Metadata Space Used: 12.72MB
 Metadata Space Total: 2.147GB
 Metadata Space Available: 2.135GB
 Thin Pool Minimum Free Space: 10.74GB
 Udev Sync Supported: true
 Deferred Removal Enabled: false
 Deferred Deletion Enabled: false
 Deferred Deleted Device Count: 0
 Data loop file: /var/lib/docker/devicemapper/devicemapper/data
 Metadata loop file: /var/lib/docker/devicemapper/devicemapper/metadata
 Library Version: 1.02.138 (2017-03-28)
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: bridge host macvlan null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: 
containerd version: 422e31ce907fd9c3833a38d7b8fdd023e5a76e73
runc version: 9c2d8d184e5da67c95d601382adf14862e4f2228
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.10.8-1-ARCH
Operating System: Arch Linux
OSType: linux
Architecture: x86_64
CPUs: 8
Total Memory: 15.6GiB
ID: JKFE:OZKT:K5ER:BR2K:4OCH:HH3Y:AZT4:P7TS:YD7I:PVVR:4VKI:CCXE
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

WARNING: devicemapper: usage of loopback devices is strongly discouraged for production use.
         Use `--storage-opt dm.thinpooldev` to specify a custom block storage device.
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

Arch Linux host, systemd version 232.  `systemd-resolved` was changed to bind to `127.0.0.53` (see https://github.com/systemd/systemd/pull/4061) specifically to avoid this kind of problem, so it seems this is a bug on the Docker side of things (also since `netcat` can bind to the port without any trouble.)

**Comment by cpuguy83** [comment]
The linked PR is only adding options to configure the systemd resolver precisely because it does indeed conflict with `0.0.0.0`.
Also looks like it got merged without dealing with the actual issue of `0.0.0.0`.

Docker is only attempting to listen on the address you specified, in this case `0.0.0.0:53` on TCP and UDP.
You could change this to listen on a specific address rather than all addresses by setting `<addr>:53:53`

**Comment by cpuguy83** [comment]
If I had to guess, netcat it's because netcat uses `SO_REUSEPORT` and `SO_REUSEADDR`
```
setsockopt(3, SOL_SOCKET, SO_REUSEADDR, [1], 4) = 0
setsockopt(3, SOL_SOCKET, SO_REUSEPORT, [1], 4) = 0
bind(3, {sa_family=AF_INET, sin_port=htons(53), sin_addr=inet_addr("0.0.0.0")}, 16) = 0
```

**Comment by timwhite** [comment]
@cpuguy83 Given that netcat can listen on 0.0.0.0:53, how hard would it be to fix docker to also be able to do it?

**Comment by cpuguy83** [comment]
If there's something is bound to a port then docker cannot bind on that port.

**Comment by kahdo** [comment]
Problem is not specific to docker. I found this thread searching around "cannot serve SSH on port 53" and I have the same environiment, an AWS hosted box with Ubuntu 18.04 which boasts systemd-resolved listening on 127.0.0.53:53

:sad: has someone been able to fix this without disabling systemd-resolved?

**Comment by Malvineous** [comment]
According to [some reports](https://unix.stackexchange.com/questions/304050/how-to-avoid-conflicts-between-dnsmasq-and-systemd-resolved), if you launch the program binding on port 53 before `systemd-resolved` then it will realise the port is in use and not bind to it.

You can also disable the port 53 bind completely in `systemd-resolved` (see previous link) which may still work, if you are using programs that use the system libraries to resolve names and don't contact DNS servers directly (e.g. most apps will still work, but certain diagnostic tools like `host` and `nslookup` will probably not work unless you manually specify another DNS server on the command line.)  This is the same as starting the other program before `systemd-resolved`, only by explicitly disabling the DNS listener, the load order doesn't matter.

Having `systemd-networkd` change its port number would also be another option but it [doesn't look like this is possible at present](https://github.com/systemd/systemd/issues/4377).

However, the default configuration only listens on UDP port 53, and SSH works on TCP, so you shouldn't be getting any problems trying to run an SSH server on the same port.  My problem was caused because I wanted UDP port 53 made available to Docker as well.  Maybe you need to confirm that the `DNSStubListener` option in `/etc/systemd/resolved.conf` is indeed at the default of `udp` and it hasn't been changed to `tcp` or `yes` (tcp+udp)?

**Comment by sam-thibault** [comment]
This is an old issue and does not seem to be specific to Docker.  I will close it as stale.
